### PR TITLE
Legalfiltering

### PR DESCRIPTION
**Pull request by voisardf**
@kalbermattenm: Banzai! let's give it a go ;) thanks for your review

### CODE REVIEW

File: crdppf/static/js/Crdppf/legalDocuments.js [hunk 1]
```diff
@@ -1,13 +1,28 @@
 Ext.namespace('Crdppf');
 
-Crdppf.filterlist = {'theme':[], 'topic' : [], 'municipalitynb': 0};
+Crdppf.filterlist = {'topic' : [], 'layers':[], 'cadastrenb': 0, 'objectids':[]};
 
 Crdppf.docfilters = function(filter) {
-
+
```
Comment (kalbermattenm) [inline crdppf/static/js/Crdppf/legalDocuments.js]: Remove white spaces

File: crdppf/static/js/Crdppf/legalDocuments.js [hunk 2]
```diff
@@ -27,27 +42,31 @@ Crdppf.docfilters = function(filter) {
...
+                    }
+                } else {
+                    return record;
                 }
             }
         } else {
             for (var i = 0; i < Crdppf.filterlist['topic'].length; i++){
             // if the topicid is in the filterlist show the corresponding documents
-             if (record.get('topicfk') == Crdppf.filterlist['topic'][i].toString()) return record;
+                if (record.get('origins').indexOf(Crdppf.filterlist['topic'][i].toString()) > -1)  return record;
```
Comment (kalbermattenm) [inline crdppf/static/js/Crdppf/legalDocuments.js]: I don't really like one line if conditions... But that might be a personal choice...

File: crdppf/static/js/Crdppf/legalDocuments.js [hunk 3]
```diff
@@ -82,37 +101,37 @@ Crdppf.legalDocuments = function() {
...
+            {name: 'origins'}, 
+            //{name: 'validationdate', type:'date'},  //date validation
+            //{name: 'modifiedby'},
+            //{name: 'status'} //for historization P:provisory, V:valid; A:archived; D:deleted      
             ]
         ), 
             listeners:{
                 load: function() {
                     Crdppf.loadingCounter += 1;
+                    //console.log(Crdppf.legalDocuments.store);
```
Comment (kalbermattenm) [inline crdppf/static/js/Crdppf/legalDocuments.js]: Please remove...

File: crdppf/static/js/Crdppf/legalDocuments.js [hunk 4]
```diff
@@ -126,81 +145,83 @@ Crdppf.legalDocuments.createView = function(labels) {
...
+        
                 '<tpl for=".">',
-                    '<tpl if="this.isLegalprovision(doctype)">',
-                        '<div style="font-size:10pt;padding:5px 15px;background-color:{[xindex % 2 === 0 ? "#FFF" : "#EEE"]}">',
-                            '<h3 class="doctitle"><a href="#" onClick="window.open(\'{documenturl}\');" target="_blank">{officialnb}</a> - {officialtitle}<span style="padding-left:15px;">Date de publication:</b> {publishedsince:date("d.m.Y")}<span></h3>',
-                            '<p class="docurl">URL:</b> <a href="#" onClick="window.open(\'{documenturl}\');" target="_blank">{documenturl}</a></p>',
-                            '<br />',
-                        '</div>',
+                    '<tpl if="this.isLegalprovision(doctype) &amp;&amp; this.isCommunal(state, municipalityname)">',
+                        '<h2 style="margin-top:10px;margin-bottom:5px;">Niveau communal</h2>',
```
Comment (kalbermattenm) [inline crdppf/static/js/Crdppf/legalDocuments.js]: There is some French in here... I don't know if you want to create a multi-language app. If yes, you should do something about it...

File: crdppf/static/js/Crdppf/map.js [hunk 5]
```diff
@@ -70,17 +70,84 @@ var setInfoControl = function setInfoControl(){
...
+        maxFeatures: 4,
         clickTolerance: 15
     });
-    
-    // define actions on feature selection
-    control.events.register("featureselected", this, function(e) {
+        
+    Crdppf.PropertySelection = function(features, labels) {
+        
+        if (store) {
```
Comment (kalbermattenm) [inline crdppf/static/js/Crdppf/map.js]: Is store a global variable ? (if yes, then it is not good....)

File: crdppf/static/js/Crdppf/map.js [hunk 6]
```diff
@@ -70,17 +70,84 @@ var setInfoControl = function setInfoControl(){
...
+            
+            var combo = new Ext.form.ComboBox({
+                store: store,
+                displayField: 'displaytxt',
+                valueField: 'index',
+                typeAhead: true,
+                mode: 'local',
+                triggerAction: 'all',
+                emptyText: labels.choosePropertyMsg,
+                selectOnFocus:true,
```
Comment (kalbermattenm) [inline crdppf/static/js/Crdppf/map.js]: missing space after `:`

File: crdppf/static/js/Crdppf/map.js [hunk 7]
```diff
@@ -70,17 +70,84 @@ var setInfoControl = function setInfoControl(){
...
+                    select: function(combo, record, index) {
+                        // the index of the selected feature is returned and used to highlight the geom
+                        property = features[index];
+                        Crdppf.featureSelection(property);
+                        Ext.getCmp('propertySelectionWindow').hide();
+                    }
+                }
+            });
+
+            if (Ext.getCmp('propertySelectionWindow')) {
```
Comment (kalbermattenm) [inline crdppf/static/js/Crdppf/map.js]: If you don't use `propertySelectionWindow` variable outside of this function, then you shouldn't use `Ext.getCmp`, but rather declare your variable locally inside the function and use the scope to hide it in the listeners.

File: crdppf/static/js/Crdppf/map.js [hunk 8]
```diff
@@ -122,11 +193,18 @@ var setInfoControl = function setInfoControl(){
                             // iterate over all features: create a node for each restriction and group them by their owning layer
                             for (j=0; j<jsonData.length; j++) {
```
Comment (kalbermattenm) [inline crdppf/static/js/Crdppf/map.js]: missing `var` here... `j` is declared globally...

File: crdppf/util/pdf_classes.py [hunk 9]
```diff
@@ -296,8 +300,45 @@ def set_wms_config(self, topicid):
...
+        getmapurl = self.crdppf_wms
+
+        if getmapurl.find('?') < 0:
+            getmapurl += '?'
+        getmapurl = getmapurl + '&'.join(['%s=%s' % (key, value) for (key, value) in params.items()])
+
+        http = httplib2.Http()
+
+        h = dict(self.request.headers)
+        if urlparse(getmapurl).hostname != 'localhost': # pragma: no cover
```
Comment (kalbermattenm) [inline crdppf/util/pdf_classes.py]: You should remove all the `# pragma: no cover`... This is only used if you do Travis checks on your code (what we don't do...)

File: crdppf/util/pdf_classes.py [hunk 10]
```diff
@@ -478,6 +523,9 @@ def get_site_map(self):
         </sld:NamedLayer>
         </sld:StyledLayerDescriptor>"""
 
+        #~ sldfile2 = open(self.appconfig.slddir+'property_overlay.sld', 'r')
+        #~ sld2 = sldfile2.read()
+        #~ sdf
```
Comment (kalbermattenm) [inline crdppf/util/pdf_classes.py]: Are these comments left here on purpose ? If not, please remove

File: crdppf/util/pdf_classes.py [hunk 11]
```diff
@@ -555,15 +604,10 @@ def add_topic(self, topic):
...
-            'legalbases':[],
-            'legalprovisions':[],
-            #'legalprovisions':topic.legalprovisions, 
-            #'temporaryprovisions':topic.temporaryprovisions, 
             'authorityfk':topic.authorityfk,
-            'publicationdate':topic.publicationdate,
+            'publicationdate':topic.publicationdate
             }
- 
+
```
Comment (kalbermattenm) [inline crdppf/util/pdf_classes.py]: white spaces...

File: crdppf/util/pdf_classes.py [hunk 12]
```diff
@@ -571,13 +615,17 @@ def add_topic(self, topic):
...
-                    'layerlegalbases':None,
-                    'layerreference':None,
-                    'layerprovisions':None,
                     'features':None
                     }
+                # intersects a given layer with the feature and adds the results to the topiclist- see method add_layer
                 self.add_layer(layer)
             self.get_topic_map(topic.layers,topic.topicid)
+            # Get the list of documents related to a topic with layers and results
+            if self.topiclist[str(layer.topicfk)]['categorie']== 3:
```
Comment (kalbermattenm) [inline crdppf/util/pdf_classes.py]: missing space before `==`

File: crdppf/views/entry.py [hunk 13]
```diff
@@ -12,7 +13,27 @@ def __init__(self, request):
         
     @view_config(route_name='home', renderer = 'crdppf:templates/derived/crdppf.mako')
     def home(self):
+
         d = {
             'debug': self.debug
-        }
+            }
+
+    #~ try:
```
Comment (kalbermattenm) [inline crdppf/views/entry.py]: Shouldn't that be removed ?

File: crdppf/views/entry.py [hunk 14]
```diff
@@ -12,7 +13,27 @@ def __init__(self, request):
...
+                    'tile_format': tile.tile_format
+                    }
+        else:
+            d['plan_cadastral'] = {
+                'tile_date':'plan_cadastral_c2c',
+                'tile_format':'image/png'}
+            d['plan_ville'] = {'tile_date':'plan_ville_c2c',
+                'tile_format':'image/png'}
+    #~ except:
+        #~ pass
```
Comment (kalbermattenm) [inline crdppf/views/entry.py]: And Shouldn't these two lines also be removed ??

File: crdppf/views/entry.py [hunk 15]
```diff
@@ -12,7 +13,27 @@ def __init__(self, request):
...
-        }
+            }
+
+    #~ try:
+        from crdppfportal.models import Tiles
+        tiles = DBSession.query(Tiles).all()
+        if len(tiles) > 0:
+            for tile in tiles:
+                d[str(tile.baselayer)] = {
+                    'tile_date':tile.baselayer+u'_'+tile.tile_date,
```
Comment (kalbermattenm) [inline crdppf/views/entry.py]: Missing space after `:`

File: crdppf/views/legal_documents.py [hunk 16]
```diff
@@ -107,111 +107,134 @@ def createNewDocEntry(request):
...
+    
+    if param == 'topic':
+        legalbases = DBSession.query(LegalBases).filter_by(topic=params[param]).all()
+    else:
+        sdf
+        legalbases = DBSession.query(LegalBases).order_by(LegalBases.legalbaseid.asc()).all()
+
     for legalbase in legalbases :
-        doclist.append({
+        doclist.append({legalbase.legalbaseid:{
```
Comment (kalbermattenm) [inline crdppf/views/legal_documents.py]: Please begin the append content on a new line and add a space after each `:` in the append

File: crdppf/views/legal_documents.py [hunk 17]
```diff
@@ -107,111 +107,134 @@ def createNewDocEntry(request):
...
+def getDocumentReferences(docfilters):
+    """Gets all the id's of the documents referenced by an object, layer, topic or document.
+    """
+    referenceslist = set()
+    rereferenceslist = set()
+    referencedocs = []
+    
+    if len(docfilters) > 0:
+        for filtercriteria in docfilters:
+            references = DBSession.query(OriginReference).filter_by(fkobj = filtercriteria).all()
```
Comment (kalbermattenm) [inline crdppf/views/legal_documents.py]: no spaces before and after `=`

File: crdppf/views/legal_documents.py [hunk 18]
```diff
@@ -107,111 +107,134 @@ def createNewDocEntry(request):
...
+            documents = documents.filter(or_(LegalDocuments.chmunicipalitynb==filters['chmunicipalitynb'], LegalDocuments.chmunicipalitynb==None))
+    else:
+        documents = DBSession.query(LegalDocuments)
+    documents = documents.order_by(LegalDocuments.docid.asc()).all()
+    
+    for document in documents :
+        origins = []
+        for origin in document.origins:
+            origins.append(origin.fkobj)
         doclist.append({
```
Comment (kalbermattenm) [inline crdppf/views/legal_documents.py]: Again, quite a lot missing spaces after `:` in this append

File: crdppf/views/legal_documents.py [hunk 19]
```diff
@@ -107,111 +107,134 @@ def createNewDocEntry(request):
...
+            'revisiondate':document.revisiondate.isoformat() if document.revisiondate else None,
+            'operator':document.operator,
+            'origins': origins
         })
+        
+    return {'docs': doclist}
 
-    return {'docs': doclist}
\ No newline at end of file
+
```
Comment (kalbermattenm) [inline crdppf/views/legal_documents.py]: Add an empty new line at the end of your file

**Comment by kalbermattenm** [comment]
I have done the review.

This really to big for me to do a good review... I don't really get what all the code is doing... As I already told you, you should really separate thing, like there is some stuff regarding the tiles, some stuff regarding the legal docs and so on. I also know that it was quite hard for you to split things, but if you want an efficient review, then you should do it...

Moreover, try to restrain your line length (might it be in Javascript files or Python files) to 80 characters. Otherwise, the review is really hard in GitHub. You can activate the status bar in Scite to see how many characters you have on a line.

So, I mainly checked the syntax, but I cannot review anything regarding functionality...

We should begin to implement functionnal tests... (yeah... that's quite a word...)

File: crdppf/static/js/Crdppf/legalDocuments.js (hunk 2 above)
Comment (voisardf) [inline crdppf/static/js/Crdppf/legalDocuments.js]: in fact, neither do I. =:0)

File: crdppf/static/js/Crdppf/legalDocuments.js (hunk 4 above)
Comment (voisardf) [inline crdppf/static/js/Crdppf/legalDocuments.js]: this is bas indeed. fixed.

Comment (voisardf) [inline crdppf/static/js/Crdppf/legalDocuments.js]: bad.

File: crdppf/static/js/Crdppf/map.js (hunk 5 above)
Comment (voisardf) [inline crdppf/static/js/Crdppf/map.js]: nope, is local. line 93 :)

File: crdppf/static/js/Crdppf/map.js (hunk 8 above)
Comment (voisardf) [inline crdppf/static/js/Crdppf/map.js]: yup; and wasn't the only one. fixed

File: crdppf/util/pdf_classes.py (hunk 9 above)
Comment (voisardf) [inline crdppf/util/pdf_classes.py]: fine by me. don't even now what travis is yet...

File: crdppf/util/pdf_classes.py (hunk 10 above)
Comment (voisardf) [inline crdppf/util/pdf_classes.py]: yes they are. unfortunately

File: crdppf/static/js/Crdppf/map.js (hunk 7 above)
Comment (voisardf) [inline crdppf/static/js/Crdppf/map.js]: cheers.

**Comment by voisardf** [comment]
addressed comments and fixed issues.

**Comment by voisardf** [comment]
ok for test implementations too =:0)

File: crdppf/util/pdf_classes.py (hunk 10 above)
Comment (voisardf) [inline crdppf/util/pdf_classes.py]: finally I skipped them, will add them again when needed.

**Comment by voisardf** [comment]
merging

**Pull request merged**